receiver/prometheusexecreceiver/receiver.go

```diff
@@ -0,0 +1,295 @@
+func getSubprocessConfig(cfg *Config) (*subprocessmanager.SubprocessConfig, error) {
+	if cfg.SubprocessConfig.Command == "" {
+		return nil, fmt.Errorf("no command to execute entered in config file for %v", cfg.Name())
+	}
+
+	subprocessConfig := &subprocessmanager.SubprocessConfig{}
+
+	subprocessConfig.Command = cfg.SubprocessConfig.Command
+	subprocessConfig.Env = cfg.SubprocessConfig.Env
+
+	return subprocessConfig, nil
+}
+
+// extractName will return the receiver's given custom name (prometheus_exec/custom_name)
+func extractName(cfg *Config) string {
+	splitName := strings.SplitN(cfg.Name(), "/", 2)
+	if len(splitName) > 1 && splitName[1] != "" {
+		return splitName[1]
+	}
+	// fall back to the first part of the string, prometheus_exec
+	return splitName[0]
+}
+
+// manageProcess will put the process in an infinite starting loop which goes through the following steps
+// If the port is not defined by the user, one is generated and a new metrics receiver is built with the new port
+// All instances of {{port}} are replaced with the actual port (either defined by the user or generated)
+// Start the receiver if it is stopped (either first iteration or if it was previously shutdown)
+// We then start the subprocess, get its runtime and decide if the process is considered healthy or not -> the receiver is shutdown if it was deemed unhealthy
+// Finally, the wait time before the subprocess is restarted is computed and this goroutine sleeps for that amount of time, before restarting the loop from the start
+func (wrapper *prometheusExecReceiver) manageProcess(ctx context.Context, host component.Host) {
+	var (
+		newPort int = wrapper.originalPort
+
+		elapsed       time.Duration
+		crashCount    int
+		err           error
+		subprocessErr error
+	)
+
+	for {
+
+		// Generate a port if none was specified and if process had a low runtime (Receiver was stopped)
+		generatePort := wrapper.originalPort == 0 && elapsed <= healthyProcessTime
+		if generatePort {
+			newPort, err = wrapper.assignNewRandomPort(ctx)
+			if err != nil {
+				wrapper.logger.Error("assignNewRandomPort() error - killing this single process/receiver", zap.String("error", err.Error()))
+				return
+			}
+		}
+
+		wrapper.subprocessConfig = wrapper.fillPortPlaceholders(newPort)
+
+		// Start the receiver if it's the first pass, or if the process is unhealthy/a newport was generated meaning the Receiver was stopped last pass
+		firstRun := elapsed == 0
+		unhealthyProcess := elapsed <= healthyProcessTime && crashCount > healthyCrashCount
+
+		if firstRun || unhealthyProcess || generatePort {
+			err = wrapper.prometheusReceiver.Start(ctx, host)
+			if err != nil {
+				wrapper.logger.Error("Start() error, could not start receiver - killing this single process/receiver", zap.String("error", err.Error()))
+				return
+			}
+		}
+
+		// Start the process, error is handled a little later
+		elapsed, subprocessErr = wrapper.subprocessConfig.Run(wrapper.logger)
```

Comment:
Will `subprocessErr` ever be non nil and what does that case mean? Also, I would recommend logging the error here straight away. You can print another log message later with the wait time

Reply:
subprocessErr **can** be non-nil, say a process that exits on purpose - without crashing, but in practice it will mostly always be defined. Noted on the error logging!

Reply:
In that case you treat it the same as if an error had occurred though right? That's something that probably **is** worth a code comment 😄